torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -295,142 +295,191 @@ def pipeline_llama_tracer(
+    logger.info(f"Applied {ac_config.mode} activation checkpointing to the model")
+    return model
 
-    # apply DP (FSDP2)
-    if parallel_dims.dp_enabled:
-        dp_mesh = world_mesh["dp"] if world_mesh.ndim > 1 else world_mesh
-        assert dp_mesh.mesh_dim_names == ("dp",), dp_mesh.mesh_dim_names
-        mp_policy = MixedPrecisionPolicy(
-            param_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param],
-            reduce_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_reduce],
+
+def apply_compile(model, world_mesh, parallel_dims, job_config: JobConfig):
+    """
+    Apply torch.compile to the model.
+    """
+
+    assert job_config.training.compile, "torch.compile is not enabled"
+    if job_config.model.norm_type == "fused_rmsnorm":
+        raise NotImplementedError(
+            "fused_rmsnorm not yet compatible with torch.compile. Please use layernorm or rmsnorm."
         )
-        fsdp_config = {"mesh": dp_mesh, "mp_policy": mp_policy}
-        for layer_id, transformer_block in model.layers.items():
-            # As an optimization, do not reshard after forward for the last
-            # transformer block since FSDP would prefetch it immediately.
-            # When using Pipeline Parallelism, generally zero-2 is best so as to avoid repeated reshardings
-            # per microbatch.
-            reshard_after_forward = (
-                int(layer_id) < len(model.layers) - 1 and not parallel_dims.pp_enabled
-            )
-            fully_shard(
-                transformer_block,
-                **fsdp_config,
-                reshard_after_forward=reshard_after_forward,
-            )
-            model.layers[layer_id] = transformer_block
-        model = fully_shard(
-            model, **fsdp_config, reshard_after_forward=not parallel_dims.pp_enabled
+
+    for layer_id, transformer_block in model.layers.named_children():
+        # turn on per-transformer block compile after AC wrapping and before FSDP
+        # TODO: dynamic shape have some issues so we turn it off for now.
+        # TODO: inline inbuilt nn modules does not work yet, enable it to accelarate
+        # compile time.
+        # torch._dynamo.config.inline_inbuilt_nn_modules = True
+        transformer_block = torch.compile(transformer_block, dynamic=False)
+        model.layers.register_module(layer_id, transformer_block)
+
+    ac_config = job_config.activation_checkpoint
+    if ac_config.mode == "selective" and ac_config.selective_ac_option == "op":
+        # some temp flags for torch.compile enablement + SAC
+        torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = (
+            True
         )
-        logger.info("Applied FSDP to the model")
+
+    logger.info("Compiled each TransformerBlock with torch.compile")
+    return model
+
+
+def apply_dp(model, world_mesh, parallel_dims, job_config: JobConfig):
```

Comment:
ditto: pass in `dp_mesh` instead of `world_mesh`
I think passing in `job_config.training` as `training_config` is good enough.

Reply:
Yeah, I was trying it following this thought.
But this line of code made me back off:
```
dp_mesh = world_mesh["dp"] if world_mesh.ndim > 1 else world_mesh
```
Too detailed to be at the top level.

Reply:
There's a trade-off of putting details here and putting details at the "top" level. My point is that by passing in world_mesh to each individual parallelism (tp, dp), you are passing in objects which shouldn't be visible to the function itself, which violates why we design device mesh in the first place -- to hide complexity when working on a particular dimension.

If we think it's too detailed, there are other ways to mitigate, e.g. Will's suggestion.

Reply:
But this can be resolved in the proposal @wconstab mentioned above.
Likely in a different PR due to the scope.